src/core/algorithms/gfd/gfd_miner.cpp

```diff
@@ -0,0 +1,683 @@
+    }
+}
+
+template <typename T>
+std::vector<std::vector<T>> GetSubsets(std::vector<T> const& elements, int const& n) {
+    if (static_cast<int>(elements.size()) < n) {
+        return {};
+    }
+    std::vector<std::vector<T>> result = {};
+    std::vector<int> indices(n);
+    std::iota(std::begin(indices), std::end(indices), 0);
+    while (!indices.empty()) {
+        std::vector<T> element = {};
+        for (auto index : indices) {
+            element.push_back(elements.at(index));
+        }
+        result.push_back(element);
+        NextSubset(indices, elements.size());
+    }
+    return result;
+}
+
+bool Satisfied(std::vector<Literal> const& literals, graph_t const& graph, graph_t const& pattern,
+               Embedding const& embedding) {
+    for (Literal const& l : literals) {
+        auto fst_token = l.first;
+        auto snd_token = l.second;
+        std::string fst;
+        std::string snd;
+        if (fst_token.first == -1) {
+            fst = fst_token.second;
+        } else {
+            vertex_t v;
+            vertex_t u = boost::vertex(fst_token.first, pattern);
+            v = embedding.at(u);
+            auto attrs = graph[v].attributes;
+            if (attrs.find(fst_token.second) == attrs.end()) {
+                return false;
+            }
+            fst = attrs.at(fst_token.second);
+        }
+        if (snd_token.first == -1) {
+            snd = snd_token.second;
+        } else {
+            vertex_t v;
+            vertex_t u = boost::vertex(snd_token.first, pattern);
+            v = embedding.at(u);
+            auto attrs = graph[v].attributes;
+            if (attrs.find(snd_token.second) == attrs.end()) {
+                return false;
+            }
+            snd = attrs.at(snd_token.second);
+        }
+        if (fst != snd) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool Validate(graph_t const& graph, Gfd const& gfd, Embeddings const& embeddings,
+              std::size_t const& sigma) {
+    bool result = false;
+    std::map<vertex_t, std::set<vertex_t>> maps;
+    for (auto& embedding : embeddings) {
+        if (!Satisfied(gfd.GetPremises(), graph, gfd.GetPattern(), embedding)) {
+            continue;
+        }
+        result = true;
+
+        if (!Satisfied(gfd.GetConclusion(), graph, gfd.GetPattern(), embedding)) {
+            return false;
+        }
+
+        BGL_FORALL_VERTICES_T(v, gfd.GetPattern(), graph_t) {
+            if (maps.find(v) != maps.end()) {
+                maps.at(v).insert(embedding.at(v));
+            } else {
+                std::set<vertex_t> vertices = {embedding.at(v)};
+                maps.emplace(v, vertices);
+            }
+        }
+    }
+
+    std::size_t min_size = boost::num_vertices(graph);
+    for (auto& [v, image] : maps) {
+        min_size = image.size() < min_size ? image.size() : min_size;
+    }
+    return (min_size >= sigma) && result;
```

Comment:
nit: changing the order of checking `result && min_size >= sigma` is slighlty faster, because if `result` is `false` there is not need to compare numbers, which is slightly slower operation